### ISSUE TITLE
Persist widget state in all open browser tabs.

### ISSUE THREAD

**Issue by najaarv** [eclipse-theia/theia#7682]
Hi, let's say I have two terminal widgets open in a theia workspace in one browser tab(Say Chrome).
I open another Chrome tab and I find that the terminal widgets are not there. I have a clean workspace. In my use case, I want both the terminal widgets to be open in all the chrome tabs I have open. What do I need to modify in the terminal-frontend-module file to make the widgets persist across tabs/sessions? 

Do I have to change some logic here:
https://github.com/eclipse-theia/theia/blob/master/packages/terminal/src/browser/terminal-frontend-module.ts#L54-L75

Help wanted. Thanks.

**Comment by kittaakos** [comment]
@najaarv, someone can correct me if I'm wrong, but the problem is that the layout persisted only before the document unloads:

https://github.com/eclipse-theia/theia/blob/498022dc88d929a10a1f973c831b4372a08a77b3/packages/core/src/browser/frontend-application.ts#L186-L190

So you open your first tab, open the terminals, and so on, and when you open the second tab with the same URL, your previous tab state has not been persisted, so the second cannot restore it.

**Comment by najaarv** [comment]
@kittaakos Thanks for pointing that out. Can I modify something on the widget level so that the widget does not have to wait until `beforeunload` event to save its state? Can you suggest any workaround? Btw yeah, the widget does implement `StatefulWidget` nterface and has `storeState` and `restoreState` methods. Thanks.

**Comment by kittaakos** [comment]
Do you have your own Theia extension? Can you add custom implementation and rebind it?

**Comment by najaarv** [comment]
Yes, I have my own Theia extension and I can add a custom implementation. Can you tell me what I need to change to make my use-case work?

**Comment by kittaakos** [comment]
> Yes, I have my own Theia extension and I can add a custom implementation.

Super 👍 

> Can you tell me what I need to change to make my use-case work?

I tried it locally, and it works for the terminal as you would expect.
```js
import { injectable, ContainerModule } from 'inversify';
import { ShellLayoutRestorer } from '@theia/core/lib/browser/shell/shell-layout-restorer';
import { FrontendApplicationContribution, FrontendApplication } from '@theia/core/lib/browser/frontend-application';

export default new ContainerModule((bind, unbind, isBound, rebind) => {
    bind(CustomShellLayoutRestorer).toSelf().inSingletonScope();
    rebind(ShellLayoutRestorer).toService(CustomShellLayoutRestorer);
    bind(FrontendApplicationContribution).toService(CustomShellLayoutRestorer);
});

@injectable()
export class CustomShellLayoutRestorer extends ShellLayoutRestorer implements FrontendApplicationContribution {

    onStart(app: FrontendApplication): void {
        app.shell.onDidChangeActiveWidget(({ newValue }) => {
            console.log(`Active widget changed: ${newValue ? newValue.id : 'N/A'}`);
            this.storeLayout(app);
        });
    }

    storeLayout(app: FrontendApplication): void {
        console.log('storing shell layout');
        super.storeLayout(app);
    }

}
```

I have not had a better idea now, I wanted to inject `ShellLayoutRestorer` to the `ApplicationShell`, but it resulted in a cycle.

So we are tracing the `active` widgets, and when there is a new, we save the layout. You can track the `current` widgets too. I hope this helps, if so, please close the issue 😊 Thanks!

**Comment by akosyakov** [comment]
Is it a request to persist layout state when the window gets invisible, so if open another tab it loads such state? Ot it is just a question?

**Comment by akosyakov** [comment]
I would expect that one uses https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event to implement it.

**Comment by najaarv** [comment]
@kittaakos Thanks very much. I'm yet to implement and fully test it, but thanks for the help. I really appreciate it. 

@akosyakov, This is a request actually. and as I mentioned in my earlier comment, I feel this feature has to be implemented for the entire Application because users would prefer to have their workspaces consistent across all browser tabs they have open.

**Comment by akosyakov** [comment]
> feature has to be implemented for the entire Application because users would prefer to have their workspaces consistent across all browser tabs they have open.

What is a use case to have the same window opened twice? It is not clear what consistent means, a user can open 2 windows for the same origin and change them differently, then they are inconsistent. The last one wins. We can do it a bit better to a new window by keeping them consistent all the time is not feasible, not sure whether it is actually desirable.

**Comment by najaarv** [comment]
> We can do it a bit better to a new window, but keeping them consistent all the time is not feasible, not sure whether it is actually desirable.

I get your point, and it makes sense. Yeah, It would be better to give the user the latest widget state when he opens a new tab/window, and then he can build on it and the latest one wins. This feature would be a really nice one to have. You can close this issue if needed, or keep it open if you plan on working on it, thanks.

**Comment by akosyakov** [comment]
@najaarv I don't mind to implement it, but i will need to justify it. If you can share why you want to work with 2 windows for the same project  in the same browser it would help.

For instance there is also such issue: https://github.com/eclipse-theia/theia/issues/526 Is it something among these lines as well?

**Comment by najaarv** [comment]
Alright, here's a scenario on the top of my mind: Say, a user has several tabs open in his browser. He needs to continue working on his theia workspace but he cannot find the tab in which he has set up his workspace, or is rather lazy to find it amongst the cluster of tabs he has left open. So he opens a new browser window or a tab, types down the URL, and would expect to continue to work from where he left off. So, ideally, he would expect the latest changes in the new browser tab as well. There you go.

**Comment by jnorwood** [comment]
I suspect everyone here is aware of the earlier eclipse e4, which kept track of configuration of multiple windows on multiple screens in the single Application.e4xmi file when running as a desktop app.    

I just tested out an e4 RAP example and modified it to have multiple trimmed windows.  The trimmed windows display within a single browser, but I suppose if you expanded the browser across multiple screens, you could drag the trimmed windows to different screens.  Is it acceptable to do something like this?

![Application e4xmi](https://user-images.githubusercontent.com/1499591/125153974-1787a980-e11d-11eb-88a1-69f79e747561.png)
![e4RAPBrowser](https://user-images.githubusercontent.com/1499591/125153980-1d7d8a80-e11d-11eb-9550-c6c2617a215e.png)
